### PR TITLE
change the validation logic with regression  validation logic, 

### PR DESCRIPTION
**Pull request by Eyobyb**
# Description
change the validation process with the regression  validation which is explained here https://linear.app/aggregate-intellect/issue/SHX-59

> Add directions for contributors. For example, do developers need to run `poetry install` after picking up these changes?

# Type of change
- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Maintenance
- [ ] New release

#SHX-59

# Checklists
To speed up the review process, please follow these checklists:

## Development
- [X] The Pull Request is small and focused on one topic
- [X] Lint rules pass locally (`make format && make lint`)
- [X] The code changed/added as part of this pull request has been covered with tests
- [X] All tests related to the changed code pass in development (`make test`)
- [ ] The changes generate no new warnings (or explain any new warnings and why they're ok)
- [X] Commit messages are detailed
- [ ] Changed code is self-explanatory and/or I added comments
- [ ] I updated the documentation (docstrings, /docs)
See the testing guidelines for help on tests, especially those involving web services.

## Code review
- [X] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached.
- [X] I have performed a self-review of my code
- [X] Issue from task tracker has a link to this pull request

💔 Thank you for submitting a pull request!

### CODE REVIEW

File: src/sherpa_ai/agents/base.py [hunk 1]
```diff
@@ -119,35 +125,85 @@ def validate_output(self):
...
-        failed_validation = []
-        result = self.synthesize_output()
+        result = ""
+        # create array of instance of validation so that we can keep track of how many times regeneration happened.
+        instantiated_validations = []
         for validation in self.validations:
-            for count in range(self.validation_steps):
-                self.belief.update_internal(EventType.result, self.name, result)
+            try:
+                instantiated_validations.append(validation())
```
Comment (20001LastOrder) [inline src/sherpa_ai/agents/base.py:133]: Can a validation be callable?

File: src/sherpa_ai/agents/base.py [hunk 2]
```diff
@@ -119,35 +125,85 @@ def validate_output(self):
...
+        all_pass = False
+        validation_is_scaped = False
+        while_count = 0
+
+        # this loop will run until max regeneration reached or all validations have failed
+        while (
+            self.global_regen_max > sum(val.count for val in instantiated_validations)
+            and not all_pass
+        ):
+            logger.info(f"validations_size: {len(instantiated_validations)}")
```
Comment (20001LastOrder) [inline src/sherpa_ai/agents/base.py:146]: This part is getting over complicated, we should extract one iteration of validation into another function

File: src/sherpa_ai/agents/base.py [hunk 3]
```diff
@@ -119,35 +125,85 @@ def validate_output(self):
...
+            break_while = False
+            if break_while:
+                break
```
Comment (20001LastOrder) [inline src/sherpa_ai/agents/base.py:149]: This is meaningless code

File: src/sherpa_ai/agents/base.py [hunk 4]
```diff
@@ -119,35 +125,85 @@ def validate_output(self):
...
-            for count in range(self.validation_steps):
-                self.belief.update_internal(EventType.result, self.name, result)
+            try:
+                instantiated_validations.append(validation())
+            except Exception as e:
+                instantiated_validations.append(validation)
+
+        all_pass = False
+        validation_is_scaped = False
+        while_count = 0
```
Comment (20001LastOrder) [inline src/sherpa_ai/agents/base.py:139]: `iteration_count`

File: src/sherpa_ai/agents/base.py [hunk 5]
```diff
@@ -119,35 +125,85 @@ def validate_output(self):
...
+            and not all_pass
+        ):
+            logger.info(f"validations_size: {len(instantiated_validations)}")
+            break_while = False
+            if break_while:
+                break
+            while_count += 1
+            logger.info(f"main_iteration: {while_count}")
+            logger.info(
+                f"regen_count: {sum(val.count for val in instantiated_validations)}"
```
Comment (20001LastOrder) [inline src/sherpa_ai/agents/base.py:153]: You can keep track of the global number of regenerations so you don't need to sum this everytime

File: src/sherpa_ai/agents/base.py [hunk 6]
```diff
@@ -119,35 +125,85 @@ def validate_output(self):
...
+            )
+            for x in range(len(instantiated_validations)):
```
Comment (20001LastOrder) [inline src/sherpa_ai/agents/base.py:155]: I would prefer `i` for loop iterator

File: src/sherpa_ai/agents/base.py [hunk 7]
```diff
@@ -119,35 +125,85 @@ def validate_output(self):
...
+                validation = instantiated_validations[x]
+                logger.info(f"validation_running: {validation.__class__.__name__}")
+                logger.info(f"validation_count: {validation.count}")
+                if validation.count < self.validation_steps:
```
Comment (20001LastOrder) [inline src/sherpa_ai/agents/base.py:159]: Shouldn't the `validation_steps` associated with each validation?

File: src/sherpa_ai/agents/base.py [hunk 8]
```diff
@@ -119,35 +125,85 @@ def validate_output(self):
...
+                        self.belief.update_internal(
+                            EventType.feedback,
+                            self.feedback_agent_name,
+                            validation_result.feedback,
+                        )
+                        break
+                    elif x == len(instantiated_validations) - 1:
+                        all_pass = True
+                elif x == len(instantiated_validations) - 1:
+                    validation_is_scaped = True
```
Comment (20001LastOrder) [inline src/sherpa_ai/agents/base.py:176]: What does this mean?

File: src/sherpa_ai/agents/base.py [hunk 9]
```diff
@@ -119,35 +125,85 @@ def validate_output(self):
...
-        if len(failed_validation) > 0:
-            # if the validation failed after all steps, append the error messages to the result
             result += "\n".join(
                 failed_validation.get_failure_message()
-                for failed_validation in failed_validation
+                for failed_validation in failed_validations
+            )
+
+        else:
+            result += "\n".join(
```
Comment (20001LastOrder) [inline src/sherpa_ai/agents/base.py:200]: I don't understand the purpose of this branch. Why do we include failure messages from all instantiated validations?

**Review by 20001LastOrder — COMMENTED** [review]
Thanks for the PR, please check the comments below as well as the issue about regenerated text.

File: src/sherpa_ai/agents/base.py (hunk 7 above)
Comment (Eyobyb) [reply src/sherpa_ai/agents/base.py:162]: it does the validation.count is the number of times it run

File: src/sherpa_ai/agents/base.py (hunk 1 above)
Comment (20001LastOrder) [reply src/sherpa_ai/agents/base.py:135]: @Eyobyb Could you help me understand this part?

File: src/sherpa_ai/agents/base.py (hunk 7 above)
Comment (20001LastOrder) [reply src/sherpa_ai/agents/base.py:154]: Yeah, this part makes sense. I'm thinking that in the design, we ideally want to be able to set a bound on the validation steps differently for each validator, isn't this the case here?

**Review by 20001LastOrder — APPROVED** [review]
LGTM, well done!

**Pull request merged**